Retry logic for getBlobsV2 in peerDAS

satushh opened the pull request:
**What type of PR is this?**
Feature

**What does this PR do? Why is it needed?**
Before this PR, the engine API method `getBlobsV2` was called only once, when the block is received from gossip.

After this PR, he engined API method `getBlobsV2` is called:
- When the block is received from gossip, or
- When a data column sidecar is received from gossip.

Additionally,  this engine API method is called periodically until the beacon node retrieved all sidecars is needs (regardless how it retrieved them).

When some data columns sidecars are constructed from the EL, the following log is now displayed:

```
[2025-09-12 16:54:49.83] DEBUG sync: Data columns sidecars constructed from the execution client via a beacon block indices=[75 86 114 117 124] iteration=0 root=0xcde999b12beb82eab37eddd71aff33b32102c8c4d5edec8977224e1ec996024d slot=367324
```

or

```
[2025-09-12 16:55:12.85] DEBUG sync: Data columns sidecars constructed from the execution client via a data column sidecar indices=[75 124] iteration=0 root=0xd64a875f3f57ad91fbe7f3e37da89db9d31783b30e7fc026a0cfd7a476571e90 slot=367326
```

Also, this PR changes some functions initially using a protobuf signed beacon block into a `ROBlock`, which is more convenient to use.

**Acknowledgements**
- [x] I have read [CONTRIBUTING.md](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md).
- [x] I have included a uniquely named [changelog fragment file](https://github.com/prysmaticlabs/prysm/blob/develop/CONTRIBUTING.md#maintaining-changelogmd).
- [x] I have added a description to this PR with sufficient context for reviewers to understand this PR.

satushh committed:
getBlobsV2: retry if reconstruction isnt successful

satushh committed:
test: engine client and sync package, metrics

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/OffchainLabs/prysm?pullRequest=15520) <br/>All committers have signed the CLA.

satushh committed:
lint: fmt and log capitalisation

satushh committed:
lint: return error when it is not nil

satushh committed:
config: make retry interval configurable

satushh committed:
sidecar: recover function and different context for retrying

satushh committed:
lint: remove unused field

satushh committed:
beacon: default retry interval

james-prysm commented on api/client/builder/client_test.go:
```diff
@@ -1727,7 +1727,7 @@ func TestSubmitBlindedBlock_BlobsBundlerInterface(t *testing.T) {
 	t.Run("Interface signature verification", func(t *testing.T) {
 		// This test verifies that the SubmitBlindedBlock method signature
 		// has been updated to return BlobsBundler interface
-		
```
please revert this if the file isn't touched

nalepae replied:
It removes a trailing space.

james-prysm commented on beacon-chain/blockchain/service.go:
```diff
@@ -588,6 +588,20 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
 
+// IsDataAvailable implements the DataAvailabilityChecker interface for use by the execution service.
+// It checks if all required blob and data column data is available for the given block.
+func (s *Service) IsDataAvailable(ctx context.Context, blockRoot [32]byte, signedBlock interfaces.ReadOnlySignedBeaconBlock) (bool, error) {
```
hmm wondering do we actually need this function? can't we just interpret err == nil as available with the existing function? or make that public? ( wondering if there's more benefits to this way of checking)

satushh replied:
I agree. 
But after checking a bit it seems that the `isDataAvailable()` **blocks** until all sidecars are available. So that can't be used to do checks in this case. I was wrong assuming before that it simply gives a yes/no answer. 
As a simple change I created extra functions (`areDataColumnsImmediatelyAvailable`, `areBlobsImmediatelyAvailable`) which only do the checks and don't wait for everything to be available. 
Please correct me if I am wrong.

nalepae replied:
Usually, we don't return a bool and an error, but just an error.
Here, you can create a new error type instead of returning true, then the caller will check the error type with `errors.Is` and decide what to do.

nalepae replied:
Please use `fieldparams.RootLength` instead of `32`.

james-prysm commented on beacon-chain/execution/engine_client.go:
```diff
@@ -1013,6 +1080,79 @@ func toBlockNumArg(number *big.Int) string {
+			attemptCount++
+
+			// Check if data is now available from any source
+			if s.isDataAlreadyAvailable(retryCtx, blockRoot, signedROBlock) {
```
if IsDataAvailable errors this will be false and it will try to reconstruct with data that may be available but was erroring for some reason

james-prysm commented on beacon-chain/sync/subscriber_data_column_sidecar.go:
```diff
@@ -29,6 +31,12 @@ func (s *Service) dataColumnSubscriber(ctx context.Context, msg proto.Message) e
+	// Trigger getBlobsV2 when receiving data column sidecar
+	if err := s.triggerGetBlobsV2ForDataColumnSidecar(ctx, root); err != nil {
+		// Log error but don't fail - this is a best-effort retry trigger
+		log.WithError(err).Debug("Failed to trigger getBlobsV2 for data column sidecar")
```
hmm will re read this part, but I feel like we should be giving a special type of error if we want to retry, arent' there errors where we wouldn't want to retry on?

james-prysm commented on proto/engine/v1/json_marshal_unmarshal_test.go:
```diff
@@ -631,7 +631,7 @@ func TestJsonMarshalUnmarshal(t *testing.T) {
 			ExcessBlobGas: 2048,
 		}
-		
+
```
lets revert this if not needed

nalepae replied:
This is actually a good change. It removes a trailing whitespace. (Not sure how this trailing whitespace landed here however...)

james-prysm commented on tools/analyzers/logcapitalization/analyzer_test.go:
```diff
@@ -18,4 +18,4 @@ func init() {
 	analysistest.RunWithSuggestedFixes(t, testdata, logcapitalization.Analyzer, "a")
-}
\ No newline at end of file
+}
```
lets revert this if not needed

nalepae replied:
It is a good change. It adds a trailing new line. (Not sure how the trailing new line was not here already before however.)

james-prysm commented on beacon-chain/sync/subscriber_data_column_sidecar.go:
```diff
@@ -52,3 +60,68 @@ func (s *Service) receiveDataColumnSidecar(ctx context.Context, sidecar blocks.V
+// This function attempts to fetch the block and trigger the execution service's retry mechanism.
+func (s *Service) triggerGetBlobsV2ForDataColumnSidecar(ctx context.Context, blockRoot [32]byte) error {
+	// Check if service is properly configured
+	if s.cfg == nil || s.cfg.chain == nil || s.cfg.beaconDB == nil {
```
I don't think we need to do this, you can check some other functions but some we just safely assume it's set

james-prysm commented on beacon-chain/sync/subscriber_data_column_sidecar.go:
```diff
@@ -52,3 +60,68 @@ func (s *Service) receiveDataColumnSidecar(ctx context.Context, sidecar blocks.V
+		}()
+
+		bgCtx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+		defer cancel()
```
I think defer is LIFO so wouldn't this cancel before the recover?

james-prysm commented on beacon-chain/sync/subscriber_data_column_sidecar.go:
```diff
@@ -52,3 +60,68 @@ func (s *Service) receiveDataColumnSidecar(ctx context.Context, sidecar blocks.V
+		if s.cfg.executionReconstructor != nil {
+			_, err := s.cfg.executionReconstructor.ReconstructDataColumnSidecars(bgCtx, signedBlock, blockRoot)
+			if err != nil {
+				log.WithError(err).Debug("getBlobsV2 retry triggered by data column sidecar failed")
```
hmm this shouldn't be debug right

nalepae commented on cmd/beacon-chain/flags/base.go:
```diff
@@ -257,6 +257,13 @@ var (
 	}
+	// GetBlobsRetryIntervalMs defines the retry interval in milliseconds for getBlobsV2 calls when reconstructing data column sidecars.
+	// Default value matches defaultGetBlobsRetryInterval in beacon-chain/execution/engine_client.go.
+	GetBlobsRetryIntervalMs = &cli.Uint64Flag{
```
I don't think we should set this as an user flag. Let's keep it as a harcoded alue.

nalepae commented on cmd/beacon-chain/execution/options.go:
```diff
@@ -23,10 +24,12 @@ func FlagOptions(c *cli.Context) ([]execution.Option, error) {
 		execution.WithHttpEndpoint(endpoint),
 		execution.WithEth1HeaderRequestLimit(c.Uint64(flags.Eth1HeaderReqLimit.Name)),
 		execution.WithHeaders(headers),
+		execution.WithGetBlobsRetryInterval(retryInterval),
```
I think this retry interval could be defined as a hard coded const in the code. No need to set it as an option nor a service field.

nalepae commented on beacon-chain/sync/subscriber_data_column_sidecar.go:
```diff
@@ -52,3 +60,68 @@ func (s *Service) receiveDataColumnSidecar(ctx context.Context, sidecar blocks.V
+	}
+
+	// Try to get the block from the database or cache
+	if !s.cfg.chain.HasBlock(ctx, blockRoot) {
```
What's the whole point to call `getBlobsV2` only if we have the block? We already to that when we receive the block.
The whole point it to call `getBlobsV2` when we receive a data column sidecar via gossip.

nalepae commented on beacon-chain/sync/subscriber_data_column_sidecar.go:
```diff
@@ -52,3 +58,54 @@ func (s *Service) receiveDataColumnSidecar(ctx context.Context, sidecar blocks.V
+	}
+
+	// Trigger the retry by calling the execution service's reconstruct method
+	go func() {
```
Why running it in a goroutine?
Also what happens if we receive all data column sidecars at about the same time. We will call up to 128 times `getBlobsV2` at about the same time?

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -659,7 +673,60 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
 func (s *Service) ReconstructDataColumnSidecars(ctx context.Context, signedROBlock interfaces.ReadOnlySignedBeaconBlock, blockRoot [fieldparams.RootLength]byte) ([]blocks.VerifiedRODataColumn, error) {
+	// Check if there's already an in-flight reconstruction call for this block
+	if resultChanInterface, exists := s.activeReconstructCalls.Load(blockRoot); exists {
+		resultChan := resultChanInterface.(chan reconstructResult)
```
use the `resultChan, ok :=` pattern to avoid panicking.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -659,7 +673,60 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+	resultChan := make(chan reconstructResult, 1)
+
+	// Try to claim this reconstruction call
+	if existingChanInterface, loaded := s.activeReconstructCalls.LoadOrStore(blockRoot, resultChan); loaded {
```
Could'nt we only do one `LoadOrStore` instead a `Load` then a `LoadOrStore`?

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -659,7 +673,60 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+
+	// We claimed the reconstruction call, ensure cleanup
+	defer s.activeReconstructCalls.Delete(blockRoot)
+	defer close(resultChan)
```
Should'nt this `defer close` action be done directly after `resultChan := make`?

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -659,7 +673,60 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+	defer close(resultChan)
+
+	// Perform the actual reconstruction
+	result, err := s.reconstructDataColumnSidecarsOnce(ctx, signedROBlock, blockRoot)
```
If err is not nil, should return with the wrapped error.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -1013,6 +1080,79 @@ func toBlockNumArg(number *big.Int) string {
+}
+
+// isDataAlreadyAvailable checks if all required data is already available for the given block.
+func (s *Service) isDataAlreadyAvailable(ctx context.Context, blockRoot [fieldparams.RootLength]byte, signedROBlock interfaces.ReadOnlySignedBeaconBlock) bool {
```
No need to create this function.
Don't log errors. Wrap them and let decide the caller what to do with.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -659,7 +673,60 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+	// Only start retry if we need more data and don't have an active retry
+	if (err != nil || len(result) == 0) && !s.hasActiveRetry(blockRoot) {
+		// Check if we already have all required data
+		if !s.isDataAlreadyAvailable(ctx, blockRoot, signedROBlock) {
```
There is a race condition here. `s.isDataAlreadyAvailable` could return `false`, then be `true` before `s.retryReconstructDataColumnSidecars` is called.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -659,7 +673,60 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+		// Check if we already have all required data
+		if !s.isDataAlreadyAvailable(ctx, blockRoot, signedROBlock) {
+			// Start background retry goroutine
+			go s.retryReconstructDataColumnSidecars(ctx, signedROBlock, blockRoot)
```
I think we don't need to define `s.reconstructDataColumnSidecarsOnce` AND `s.retryReconstructDataColumnSidecars`.
The same function could do the same. (First try to reconstruct, then if not available, then setup the ticker.)

nalepae commented on beacon-chain/sync/subscriber_data_column_sidecar.go:
```diff
@@ -29,6 +31,12 @@ func (s *Service) dataColumnSubscriber(ctx context.Context, msg proto.Message) e
 
+	// Trigger getBlobsV2 when receiving data column sidecar
+	if err := s.triggerGetBlobsV2ForDataColumnSidecar(ctx, root); err != nil {
+		// Log error but don't fail - this is a best-effort retry trigger
```
Don't really agree. Please wrap and return the error and let the caller to decide.

nalepae commented on beacon-chain/sync/subscriber_data_column_sidecar.go:
```diff
@@ -52,3 +60,68 @@ func (s *Service) receiveDataColumnSidecar(ctx context.Context, sidecar blocks.V
+		log.WithField("blockRoot", fmt.Sprintf("%#x", blockRoot)).Debug("Triggering getBlobsV2 retry for data column sidecar")
+
+		// This will trigger retry logic if needed
+		if s.cfg.executionReconstructor != nil {
```
Better to do the opposite.
Return if nil.
The continue the normal execution if not nil.

nalepae reviewed (commented):
I know this is not stricly the scope of this PR, but I think we could avoid the call to `s.cfg.executionReconstructor.ReconstructDataColumnSidecars` in `processDataColumnSidecarsFromExecution` if we already have what we need. Let's avoid to stress for free the execution client.

nalepae reviewed (commented):
Your design only retry calling GetBlobV2s when one data column sidecar is received. Not when a block is received.

nalepae commented:
General comments:

- `getBlobsV2` should be called regularly as soon as we receive a block or a data columns sidecar (first of them), until we have what we want.
- No call to `getBlobsV2` should be done if we already have what we want.
- I think a honest simplification could be to say "allow only one `getBlobsV2` call at a time, even for different blocks". It will remove the need of syncMaps etc... And just allow to use a single mutex. Like here:

https://github.com/OffchainLabs/prysm/blob/cd6cc76d5826f4a4ff880a1b8a9fa3d58772928a/beacon-chain/sync/data_columns_reconstruct.go#L36-L38

Also, it will prevent too much stress on the EL.

satushh committed:
reconstruct: load once, correctly deliver the result to all waiting goroutines

satushh committed:
reconstruct: simplify multi goroutine case and avoid race condition

satushh committed:
engine: remove isDataAlreadyAvailable function

satushh committed:
sync: no goroutine, getblobsv2 in absence of block as well, wrap error

satushh committed:
exec: hardcode retry interval

satushh committed:
da: non blocking checks

satushh committed:
sync: remove unwanted checks

satushh committed:
execution: fix test

satushh committed:
execution: retry atomicity test

satushh committed:
da: updated IsDataAvailable

satushh committed:
sync: remove unwanted tests

satushh committed:
bazel: bazel run //:gazelle -- fix

satushh committed:
blockchain: fix CustodyGroupCount return

satushh committed:
lint: formatting

satushh committed:
lint: lint and use unused metrics

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -138,8 +142,17 @@ type EngineCaller interface {
 var ErrEmptyBlockHash = errors.New("Block hash is empty 0x0000...")
 
+// Global singleflight group for data column reconstruction
+var reconstructSingleflight singleflight.Group
```
I think this should be a field of the service, not a global package variable.

satushh committed:
execution: retry logic inside ReconstructDataColumnSidecars itself

satushh committed:
lint: format

satushh committed:
execution: ensure the retry actually happens when it needs to

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -592,6 +593,32 @@ func (s *Service) updateCustodyInfoInDB(slot primitives.Slot) (primitives.Slot,
+	var err error
+
+	if blockVersion >= version.Fulu {
+		err = s.areDataColumnsImmediatelyAvailable(ctx, blockRoot, block)
```
Avoid using if/else if.
You can just use if, then directly return in the if.
In implement the code in the else if just after the if.

nalepae replied:
You also should consider differently err in case of failure and in case of unavailable data.
The same comment of line 598 about error handling applies.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +665,122 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+}
+
+// reconstructDataColumnSidecarsOnce performs a single attempt to reconstruct data column sidecars.
+// This is the original ReconstructDataColumnSidecars logic extracted for reuse in retry logic.
```
I don't understand this comment.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +665,122 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+
+// reconstructDataColumnSidecarsOnce performs a single attempt to reconstruct data column sidecars.
+// This is the original ReconstructDataColumnSidecars logic extracted for reuse in retry logic.
+func (s *Service) reconstructDataColumnSidecarsOnce(ctx context.Context, signedROBlock interfaces.ReadOnlySignedBeaconBlock, blockRoot [fieldparams.RootLength]byte) ([]blocks.VerifiedRODataColumn, error) {
```
Why renaming thing function?

satushh replied:
You mean why this function was created? This has had the same name since I created it.

nalepae replied:
Hum I don't remember why I asked this question exactly.
However why this function is named `reconstructDataColumnSidecarsOnce`?
I guess because because before the function `reconstructDataColumnSidecarsMultiple` was existing?

Could we rename it `reconstructDataColumnSidecars`?

nalepae replied:
OK I see, it's just the way how git it presenting diffs. I close the conversation.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +665,122 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+		}
+
+		// Empty result - check if data is already available
+		if s.availabilityChecker != nil {
```
Revert the logic. Test if the availability checker if nil and return on error if it is the case.

nalepae commented on beacon-chain/node/node.go:
```diff
@@ -857,6 +857,9 @@ func (b *BeaconNode) registerSyncService(initialSyncComplete chan struct{}, bFil
 	}
 
+	// Set the blockchain service as the data availability checker for getBlobsV2 retries
+	web3Service.SetDataAvailabilityChecker(chainService)
```
Actually this concerns me quite a bit. It smells a beginning of a conceptual cycle.
The execution service should remain an execution service, aka dealing with the execution layer.
It should not have any knowledge about data availability, blob verification, data columns verification etc...
These notions and associated logic should remain in the blockchain service, not in the execution service.

The execution service should not depend on the blockchain service.

satushh committed:
execution: ensure single responsibility, execution should not do DA check

nalepae committed:
PeerDAS: Implement sync

satushh committed:
sync: don't call ReconstructDataColumnSidecars if not required

satushh committed:
blockchain: move IsDataAvailable interface to blockchain package

satushh committed:
execution: make reconstructSingleflight part of the service struct

satushh committed:
blockchain: cleaner DA check

satushh committed:
lint: formatting and remove confusing comment

satushh committed:
sync: fix lint, test and add extra test for when data is actually not available

satushh committed:
sync: new appropriate mock service

satushh committed:
execution: edge case - delete activeRetries on success

nalepae committed:
Fix Potuz's comment.

nalepae committed:
Fix Potuz's comment.

nalepae committed:
Fix Potuz's comment.

nalepae committed:
Fix Potuz's comment.

nalepae committed:
Fix Potuz's comment.

nalepae committed:
Implement `TestFetchDataColumnSidecarsFromPeers`.

potuz commented on beacon-chain/blockchain/process_block.go:
```diff
@@ -901,6 +901,119 @@ func (s *Service) areBlobsAvailable(ctx context.Context, root [fieldparams.RootL
+// available in the database without waiting for missing ones.
+func (s *Service) areDataColumnsImmediatelyAvailable(
+	ctx context.Context,
+	root [fieldparams.RootLength]byte,
+	block interfaces.ReadOnlyBeaconBlock,
```
I assume this is called on a context we don't yet have a ROBlock?

satushh replied:
Changed it here https://github.com/OffchainLabs/prysm/commit/1bbb65dba84c0110ad6bad2decbb99e226d2a218

potuz commented on beacon-chain/blockchain/process_block.go:
```diff
@@ -901,6 +901,119 @@ func (s *Service) areBlobsAvailable(ctx context.Context, root [fieldparams.RootL
+	root [fieldparams.RootLength]byte,
+	block interfaces.ReadOnlyBeaconBlock,
+) error {
+	samplesPerSlot := params.BeaconConfig().SamplesPerSlot
```
Move this down to where it' s used.

satushh replied:
Did it https://github.com/OffchainLabs/prysm/blob/peerDAS-getBlobsV2/beacon-chain/blockchain/process_block.go#L966

potuz commented on beacon-chain/blockchain/service.go:
```diff
@@ -39,12 +39,22 @@ import (
+// Returns nil if data is available, ErrDataNotAvailable if data is not available,
+// or another error for other failures.
+type DataAvailabilityChecker interface {
+	IsDataAvailable(ctx context.Context, blockRoot [32]byte, signedBlock interfaces.ReadOnlySignedBeaconBlock) error
```
Same comment as above, can the pair `blockRoot, signedBlock`  be changed to a  `ROBlock`?

satushh replied:
Changed it here https://github.com/OffchainLabs/prysm/pull/15520/commits/1bbb65dba84c0110ad6bad2decbb99e226d2a218

potuz commented on beacon-chain/blockchain/testing/mock.go:
```diff
@@ -732,6 +732,11 @@ func (c *ChainService) TargetRootForEpoch(_ [32]byte, _ primitives.Epoch) ([32]b
 
+// IsDataAvailable implements the data availability checker interface for testing
+func (c *ChainService) IsDataAvailable(_ context.Context, _ [32]byte, _ interfaces.ReadOnlySignedBeaconBlock) error {
+	return nil
```
Do all test check without errors from this interface?

satushh replied:
Yes but I suppose I should add a case where the data is unavailable or gives another kind of error for completeness? Or did you mean something else?

potuz commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,92 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+		// Empty result - initiate retry mechanism
+
+		// Create a new context with a timeout for the retry goroutine.
+		retryCtx, cancel := context.WithTimeout(ctx, time.Duration(params.BeaconConfig().SecondsPerSlot)*time.Second)
```
Does the parent context need to be ` ctx`  or the background?

satushh replied:
Changed it to service context

potuz commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,92 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+					getBlobsRetryAttempts.WithLabelValues("attempt").Inc()
+
+					// Retry reconstruction
+					retryLog.WithField("attempt", attemptCount).Debug("Retrying data column reconstruction")
```
Is it worth getting a retry log every 200ms here? This will be spamming every node on Debug when the chain scales.

satushh replied:
Should there be an exponential backoff or you have any other suggestions? Like making it trace instead of debug?

nalepae replied:
Agree with potuz. This log is not useful.

potuz commented on beacon-chain/sync/subscriber_beacon_blocks.go:
```diff
@@ -108,6 +109,18 @@ func (s *Service) processDataColumnSidecarsFromExecution(ctx context.Context, ro
 		log.Warning("Data column storage is not enabled, skip saving data column, but continue to reconstruct and broadcast data column")
 	}
 
+	// Check if data is already available to avoid unnecessary execution client calls
```
Not part of this PR but I don' t understand the check in lines 108--110, if that configuration is not set we log something but continue as normal, it's an odd place for this log and not alter behavior at all. Are we attempting to save anyway?

satushh replied:
My guess is it might be for eventual broadcasting but @nalepae can confirm?

nalepae committed:
Implement `TestSelectPeers`.

nalepae committed:
Fix James' comment.

nalepae committed:
Fix flakiness in `TestSelectPeers`.

nalepae committed:
Revert "Fix Potuz's comment."

This reverts commit c45230b45509b989bdfbc104a4d486c3416c4781.

nalepae committed:
Revert "Fix James' comment."

This reverts commit a3f919205a0ff2c7ffb96246a130d82e0bd18b7f.

satushh committed:
execution: use service context instead of function's for retry

satushh committed:
blockchain: get variable samplesPerSlot only when required

nalepae committed:
`selectPeers`: Avoid map with key but empty value.

nalepae committed:
Fix Potuz's comment.

nalepae committed:
Add DataColumnStorage and SubscribeAllDataSubnets flag.

satushh committed:
remove redundant function and fix name

satushh committed:
fix test

satushh committed:
fix more tests

satushh committed:
put samplesPerSlot at appropriate place

satushh committed:
tidy up IsDataAvailable

satushh committed:
correct bad merge

satushh committed:
fix bad merge

satushh committed:
remove redundant flag option

nalepae committed:
Merge branch 'develop' into peerDAS-getBlobsV2

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -106,6 +116,9 @@ type Checker interface {
 var ErrMissingClockSetter = errors.New("blockchain Service initialized without a startup.ClockSetter")
 
+// ErrDataNotAvailable is returned when block data is not immediately available for processing.
+var ErrDataNotAvailable = errors.New("block data is not available")
```
Nit: Those 2 errors can be grouped in the same
```go
var (
    ErrMissingClockSetter = ...
    ErrDataNotAvailable = ...
)
```

declaration

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
 // ReconstructDataColumnSidecars reconstructs the verified data column sidecars for a given beacon block.
-// It retrieves the KZG commitments from the block body, fetches the associated blobs and cell proofs from the EL,
-// and constructs the corresponding verified read-only data column sidecars.
+// It uses singleflight to ensure only one reconstruction per blockRoot.
```
Nit:


```suggestion
// It uses singleflight to ensure only one reconstruction per block root.
```

nalepae replied:
This is a implementation detail, I think it should not be part of the godoc.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
-// and constructs the corresponding verified read-only data column sidecars.
+// It uses singleflight to ensure only one reconstruction per blockRoot.
 func (s *Service) ReconstructDataColumnSidecars(ctx context.Context, signedROBlock interfaces.ReadOnlySignedBeaconBlock, blockRoot [fieldparams.RootLength]byte) ([]blocks.VerifiedRODataColumn, error) {
+	// Use singleflight to ensure only one reconstruction per blockRoot
```
Nit

```suggestion
	// Use singleflight to ensure only one reconstruction per block root
```

nalepae commented on beacon-chain/sync/subscriber_beacon_blocks.go:
```diff
@@ -108,6 +109,23 @@ func (s *Service) processDataColumnSidecarsFromExecution(ctx context.Context, ro
+	}
+
 	// When this function is called, it's from the time when the block is received, so in almost all situations we need to get the data column from EL instead of the blob storage.
 	sidecars, err := s.cfg.executionReconstructor.ReconstructDataColumnSidecars(ctx, roSignedBlock, blockRoot)
```
Is it possible to use `roBlock` (which has a `.Root()` cheap method) instead of `roSignedBlock` and `blockRoot`?

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
-// It retrieves the KZG commitments from the block body, fetches the associated blobs and cell proofs from the EL,
-// and constructs the corresponding verified read-only data column sidecars.
+// It uses singleflight to ensure only one reconstruction per blockRoot.
 func (s *Service) ReconstructDataColumnSidecars(ctx context.Context, signedROBlock interfaces.ReadOnlySignedBeaconBlock, blockRoot [fieldparams.RootLength]byte) ([]blocks.VerifiedRODataColumn, error) {
```
If this function is called with a block with no commitment, it will loop until the context cancellation.

nalepae replied:
What happens if, between 2 retries, we received (via other media) all needed sidecars? Do we continue to try reconstructing until either it returns on success or the context is on error?
Reminder: All blocks with blobs proposed with an external builder won't have blobs flowing between EL in the mempool --> `getBlobsV2` will never return something in that case.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+			for {
+				select {
+				case <-ticker.C:
+					attemptCount++
```
A signed variable should not be `++`ed indefinitly. ==> It will overflow and go to negative values.
I know it likely won't be case, but it's preferable to use an unsigned variable here.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+						continue
+					}
+					if len(result) > 0 {
+						retryLog.WithField("attempts", attemptCount).Debug("Retry succeeded")
```
This log is not useful IMO. The `s.reconstructDataColumnSidecarsOnce` already logs on success.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+			defer ticker.Stop()
+
+			attemptCount := 0
+			retryLog := log.WithField("blockRoot", fmt.Sprintf("%#x", blockRoot))
```
I think adding block slot is also useful here for debug.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+					}
+
+				case <-retryCtx.Done():
+					retryLog.WithField("attempts", attemptCount).Debug("Retry timeout")
```
Not very informative log. What is retried?

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+
+					// Retry reconstruction
+					retryLog.WithField("attempt", attemptCount).Debug("Retrying data column reconstruction")
+					result, err := s.reconstructDataColumnSidecarsOnce(retryCtx, signedROBlock, blockRoot)
```
Where is this `result` value used?

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+		}()
+
+		// Return empty result for now; the background retry will handle it.
+		return []blocks.VerifiedRODataColumn{}, nil
```
Not sure to understand here. Why not returned the retrieved values?

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+	if err != nil {
+		return nil, err
+	}
+	return v.([]blocks.VerifiedRODataColumn), nil
```
May panic. Please use `result, ok` version instead.
https://github.com/uber-go/guide/blob/master/style.md#handle-type-assertion-failures

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -652,9 +654,94 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
+		// LoadOrStore atomically checks for an existing retry and stores
+		// a new one if none exists. This prevents a race condition.
+		// The stored value is the cancel function for the new context.
+		_, loaded := s.activeRetries.LoadOrStore(blockRoot, cancel)
```
I don't understand why we need both `singleFlight.group` AND `syncmap`.

nalepae commented on beacon-chain/execution/engine_client.go:
```diff
@@ -1008,6 +1095,12 @@ func toBlockNumArg(number *big.Int) string {
 }
 
+// hasActiveRetry checks if there's an active retry for the given block root.
+func (s *Service) hasActiveRetry(blockRoot [fieldparams.RootLength]byte) bool {
```
Unused function. (Only used in tests.)

nalepae commented on beacon-chain/sync/subscriber_beacon_blocks.go:
```diff
@@ -108,6 +109,23 @@ func (s *Service) processDataColumnSidecarsFromExecution(ctx context.Context, ro
+
 	// When this function is called, it's from the time when the block is received, so in almost all situations we need to get the data column from EL instead of the blob storage.
 	sidecars, err := s.cfg.executionReconstructor.ReconstructDataColumnSidecars(ctx, roSignedBlock, blockRoot)
 	if err != nil {
```
Talking about:

```go
	// Return early if no blobs are retrieved from the EL.
	if len(sidecars) == 0 {
		return
	}
```

Is is possible to have `len(sidecars) == 0` if `err != nil`?

nalepae commented on beacon-chain/sync/subscriber_beacon_blocks.go:
```diff
@@ -108,6 +109,23 @@ func (s *Service) processDataColumnSidecarsFromExecution(ctx context.Context, ro
+		log.WithError(err).Error("Failed to create ROBlock")
+		return
+	}
+	switch err := s.cfg.chain.IsDataAvailable(ctx, roBlock); {
```
`s.cfg.chain.IsDataAvailable` manages both blob sidecars and data column sidecars, but is used only for data column sidecars.

nalepae commented on beacon-chain/sync/subscriber_beacon_blocks.go:
```diff
@@ -108,6 +109,23 @@ func (s *Service) processDataColumnSidecarsFromExecution(ctx context.Context, ro
+		return
+	}
+	switch err := s.cfg.chain.IsDataAvailable(ctx, roBlock); {
+	case err == nil:
```
If feel a little bit inconfortable with the `processDataColumnSidecarsFromExecution` function.
- This function calls `s.cfg.chain.IsDataAvailable` which supports both blob sidecars and data column sidecars, but is used only for data column sidecars.
- The whole `peerdas.Info(nodeID, custodyGroupCount)` apparatus (and all previous functions needed to get `nodeID` and `custodyGroupCount`) are called two times. Once in `IsDataAvailable` and once in the end of the function before re-seeding.
- Unless I'm wrong, `ReconstructDataColumnSidecars` will continue trying to reconstruct every 200 ms even if we already have what we need.

I think a better design would be:
In a loop (every 200 ms):
- Check if all needed sidecars needed are available in the DB.
- If all sidecars are already available, then stop. If some sidecars are missing, try to reconstruct them. If OK, stop the loop and return them. If not, continue the loop.

nalepae commented on beacon-chain/blockchain/service.go:
```diff
@@ -37,12 +37,22 @@ import (
+// Returns nil if data is available, ErrDataNotAvailable if data is not available,
+// or another error for other failures.
+type DataAvailabilityChecker interface {
+	IsDataAvailable(ctx context.Context, roBlock blocks.ROBlock) error
```
This function is defined in the blockchain package but is only used in the sync package.
I think it should be defined in the sync package. (The sync package has all items needed to compute that).
Thus, no need to define a new interface is done in the sync package.

nalepae commented on beacon-chain/sync/subscriber_data_column_sidecar.go:
```diff
@@ -52,3 +58,59 @@ func (s *Service) receiveDataColumnSidecar(ctx context.Context, sidecar blocks.V
+
+// triggerGetBlobsV2ForDataColumnSidecar triggers getBlobsV2 retry when receiving a data column sidecar.
+// This function attempts to fetch the block and trigger the execution service's retry mechanism.
+func (s *Service) triggerGetBlobsV2ForDataColumnSidecar(ctx context.Context, blockRoot [32]byte) error {
```
What's the purpose of this function if it is unable to reconstruction sidecars from EL if the block is not yet received?
The idea is to get blobs from the EL, reconstruct needed sidecars and reseed them even if we don't have (yet) the corresponding block.

nalepae commented on beacon-chain/core/peerdas/validator.go:
```diff
@@ -28,3 +41,120 @@ func ValidatorsCustodyRequirement(state beaconState.ReadOnlyBeaconState, validat
+// dataColumnSidecars, given a signed block header and the commitments, inclusion proof, cells/proofs associated with
+// each blob in the block, assemble the sidecars which can be distributed to peers.
+// https://github.com/ethereum/consensus-specs/blob/master/specs/fulu/validator.md#get_data_column_sidecars
+func dataColumnSidecars(
```
This function was in `das_core.go` and is now moved into `validator.go` to stick with the beacon specification structure.

nalepae commented on beacon-chain/core/peerdas/das_core.go:
```diff
@@ -117,44 +109,6 @@ func ComputeColumnsForCustodyGroup(custodyGroup uint64) ([]uint64, error) {
-// The returned value contains pointers to function parameters.
-// (If the caller alterates `cellsAndProofs` afterwards, the returned value will be modified as well.)
-// https://github.com/ethereum/consensus-specs/blob/master/specs/fulu/validator.md#get_data_column_sidecars_from_block
-func DataColumnSidecars(signedBlock interfaces.ReadOnlySignedBeaconBlock, cellsAndProofs []kzg.CellsAndProofs) ([]*ethpb.DataColumnSidecar, error) {
```
This function was in `das_core.go` and is now moved into `validator.go` to stick with the beacon specification structure.

kasey committed:
refactor to deduplicate sidecar construction code

james-prysm commented on beacon-chain/execution/engine_client.go:
```diff
@@ -651,22 +654,73 @@ func (s *Service) ReconstructBlobSidecars(ctx context.Context, block interfaces.
 }
 
-// ReconstructDataColumnSidecars reconstructs the verified data column sidecars for a given beacon block.
+// ConstructDataColumnSidecarsFromBlock reconstructs the verified data column sidecars for a given beacon block.
```
Is construction just a better name instead of Reconstruction?

nalepae replied:
My intention is to differentiate two cases:
1. When we have `>= 64` data column sidecars, we are able to **reconstruct** all the missing sidecars be recomputing the missing cells (very cheap) and the missing proofs (very expensive).
2. However, when we retrieve blobs and proofs from the EL, the EL gives us all the proofs, for all the columns. The only reason it does not give us all the cells, is because cells are heavy to carry across the network, but really cheap to reconstruct. ==> We don't need to **reconstruct** (expensive) the missing data column sidecars from the data column sidecars we already have, but we only need to **construct** (cheap) data column sidecars from what the EL give us.

Let me know if it is unclear.

james-prysm commented on beacon-chain/execution/engine_client.go:
```diff
@@ -677,47 +731,43 @@ func (s *Service) ReconstructDataColumnSidecars(ctx context.Context, signedROBlo
 
-		blobs, cellProofs = append(blobs, blobsAndProofs.Blob), append(cellProofs, blobsAndProofs.KzgProofs...)
+		blobs = append(blobs, blobsAndProofs.Blob)
+		cellProofs = append(cellProofs, blobsAndProofs.KzgProofs...)
```
this part of the code reads a bit weird, it looks like we are looping to flatten blobs and kzg cell proos here appending them for blobs, cellProofs := make([][]byte, 0, len(blobAndProofV2s)), make([][]byte, 0, len(blobAndProofV2s))

but in ComputeCellsAndProofs it looks like we loop through it again break it apart and create []kzg.CellsAndProofs.

is there a preference doing it this way over calling ComputerCellsAndProofs multiple times and appending them together?

nalepae replied:
This issue comes from the fact that [engine_getPayloadV5](https://github.com/ethereum/execution-apis/blob/main/src/engine/osaka.md#response) response is a [BlobsBundleV2](https://github.com/ethereum/execution-apis/blob/main/src/engine/osaka.md#BlobsBundleV2), and the [engine_getBlobsV2](https://github.com/ethereum/execution-apis/blob/main/src/engine/osaka.md#engine_getblobsv2) response is a [BlobAndProofV2](https://github.com/ethereum/execution-apis/blob/main/src/engine/osaka.md#BlobAndProofV2).

One response is "flat", the other is not.

==> I wanted to have only one `ComputeCellsAndProofs` function.

But anyway, it's a waste of CPU. Will fix.

nalepae replied:
Fixed in dbf9d746ad381681edf9480674d5bd9ac1d94fbf.

james-prysm commented on beacon-chain/rpc/prysm/v1alpha1/validator/proposer.go:
```diff
@@ -468,26 +475,21 @@ func (vs *Server) broadcastAndReceiveBlobs(ctx context.Context, sidecars []*ethp
 func (vs *Server) broadcastAndReceiveDataColumns(
 	ctx context.Context,
-	sidecars []*ethpb.DataColumnSidecar,
+	roSidecars []blocks.RODataColumn,
```
since this is an interface do we need to check for nil ness in these functions?

nalepae replied:
`RODataColumn` is not an interface but a struct.
Thus, it cannot be nil.

james-prysm commented on beacon-chain/sync/subscriber_beacon_blocks.go:
```diff
@@ -61,104 +59,29 @@ func (s *Service) beaconBlockSubscriber(ctx context.Context, msg proto.Message)
+func (s *Service) processSidecarsFromExecutionFromBlock(ctx context.Context, roSignedBlock interfaces.ReadOnlySignedBeaconBlock) {
+	if roSignedBlock.Version() >= version.Fulu {
+		key := fmt.Sprintf("%#x", roSignedBlock.Block().StateRoot())
+		if _, err, _ := s.columnSidecarsExecSingleFlight.Do(key, func() (interface{}, error) {
```
what is the benefit of this new type singleflight.Group, the function executed looks a little unsafe but maybe because I don't understand this very much

nalepae replied:
What is unsafe exactly?

singleFlight is useful to ensure only one computation is running at a time.
See [here](https://pkg.go.dev/golang.org/x/sync/singleflight)

> Do executes and returns the results of the given function, making sure that only one execution is in-flight for a given key at a time. If a duplicate comes in, the duplicate caller waits for the original to complete and receives the same results. The return value shared indicates whether v was given to multiple callers.

james-prysm replied:
maybe i'm off i'm just seeing this function `func() (interface{}, error) {` will go look into how singleflight is used

nalepae replied:
For information the function executed under singleflight returns`nil`. We don't use at all `interface{}`. Only `error`.

james-prysm replied:
oh wait is this used so that 
-When the block is received from gossip, or
-When a data column sidecar is received from gossip.
don't conflict?

nalepae replied:
Yes exactly!
This function can be ran up to 129 times per block (when the block itself is received via gossip, and when up to 128 data column sidecars are received via gossip).

We need this function to run at most once. SingleFlight will ensure there this function does not run multiple times in parallel for the same root. (It would produce the exact same result and so it would only be a waste of CPU...)

james-prysm commented on beacon-chain/sync/subscriber_data_column_sidecar.go:
```diff
@@ -21,14 +21,21 @@ func (s *Service) dataColumnSubscriber(ctx context.Context, msg proto.Message) e
 
+	key := fmt.Sprintf("%#x", sidecar.GetSignedBlockHeader().Header.StateRoot)
+	if _, err, _ := s.columnSidecarsExecSingleFlight.Do(key, func() (interface{}, error) {
+		if err := s.processDataColumnSidecarsFromExecutionFromColumnSidecar(ctx, sidecar); err != nil {
```
does this name need to be this long? why not at least  processDataColumnSidecarsFromExecution 🤔

nalepae replied:
Fixed in the latest version.

james-prysm commented on beacon-chain/sync/data_columns_reconstruct.go:
```diff
@@ -220,3 +220,198 @@ func (s *Service) broadcastMissingDataColumnSidecars(
+	return nil
+}
+
+// columnIndicesToSample returns the indices of  data columns we should sample.
```
extra space here

nalepae replied:
Fixed in the latest version.

james-prysm commented on beacon-chain/sync/data_columns_reconstruct.go:
```diff
@@ -220,3 +220,198 @@ func (s *Service) broadcastMissingDataColumnSidecars(
+	}
+}
+
+func (s *Service) processDataColumnSidecarsFromExecutionFromColumnSidecar(ctx context.Context, sidecar blocks.VerifiedRODataColumn) error {
```
this name seems too long, do we have other processDataColumnSidecars functions?

nalepae replied:
Fixed in the latest version.

james-prysm commented on beacon-chain/core/peerdas/reconstruction.go:
```diff
@@ -256,6 +197,47 @@ func ReconstructBlobs(block blocks.ROBlock, verifiedDataColumnSidecars []blocks.
 }
 
+// ComputeCellsAndProofs computes the cells and proofs from blobs and cell proofs.
+func ComputeCellsAndProofs(blobs [][]byte, cellProofs [][]byte) ([]kzg.CellsAndProofs, error) {
```
is it more useful to have this kind of function that can compute from a flattened list or just have a function that takes (blob []byte, cellProof) and returns kzgCellsAndProofs object.

nalepae replied:
Fixed in dbf9d746ad381681edf9480674d5bd9ac1d94fbf.

If you have a better idea regarding function names, please let me know!

james-prysm commented on beacon-chain/sync/data_columns_reconstruct.go:
```diff
@@ -220,3 +220,198 @@ func (s *Service) broadcastMissingDataColumnSidecars(
+
+// processDataColumnSidecarsFromExecutionFromBlock retrieves (if available) data column sidecars data from the execution client,
+// builds corresponding sidecars, save them to the storage, and broadcasts them over P2P if necessary.
+func (s *Service) processDataColumnSidecarsFromExecutionFromBlock(ctx context.Context, roSignedBlock interfaces.ReadOnlySignedBeaconBlock) error {
```
oh i see there is fromExecutionFromblock and FromExecutionFromColumnSidecar, What if we used functional options as the argument and reduce ti to process DataColumnSidecars 🤔 not sure if you'd be willing to do something like that just to improve naming.

nalepae replied:
Fixed in the latest version.

james-prysm commented on beacon-chain/sync/data_columns_reconstruct.go:
```diff
@@ -220,3 +220,198 @@ func (s *Service) broadcastMissingDataColumnSidecars(
+			continue
+		}
+
+		sidecar := sidecars[index]
```
can this be out of bounds?

nalepae replied:
No. This is why there is the check:
```go
		if index >= sidecarCount {
			return errors.Errorf("data column index %d >= sidecar count %d - should never happen", index, sidecarCount)
		}
```

on line 347.

nalepae committed:
- Add godocs
- Rename some functions to be closer to the spec
- Add err in return of commitments

nalepae committed:
Replace mutating public method (but only internally used) `Populate` but private not mutating method `extract`.

nalepae committed:
Implement a unique `processDataColumnSidecarsFromExecution` instead 2 separate functions from block and from sidecar.

nalepae committed:
Merge branch 'develop' into peerDAS-getBlobsV2

nalepae committed:
`ReceiveBlock`: Wrap errors.

nalepae committed:
Remove useless tests.

nalepae committed:
`ConstructionPopulator`: Add tests.

nalepae committed:
Fix tests

nalepae committed:
Move functions to be consistent with blobs.

nalepae committed:
`fetchCellsAndProofsFromExecution`: Avoid useless flattening.

nalepae committed:
`processDataColumnSidecarsFromExecution`: Stop using DB cache.

nalepae committed:
Merge branch 'develop' into peerDAS-getBlobsV2

james-prysm commented on beacon-chain/sync/data_columns_reconstruct.go:
```diff
@@ -220,20 +220,15 @@ func (s *Service) broadcastMissingDataColumnSidecars(
+// processDataColumnSidecarsFromExecution retrieves (if available) data column sidecars data from the execution client,
 // builds corresponding sidecars, save them to the storage, and broadcasts them over P2P if necessary.
-func (s *Service) processDataColumnSidecarsFromExecutionFromBlock(ctx context.Context, roBlock blocks.ROBlock) error {
+func (s *Service) processDataColumnSidecarsFromExecution(ctx context.Context, source peerdas.ConstructionPopulator) error {
```
nice this reads better

james-prysm reviewed (approved):
all my concerns were addressed, and went through it twice
